### ISSUE TITLE
issue with docker-compose, error during compose (some crazy UID / GID)

### ISSUE THREAD

**Issue by anthonyattacks** [SamuelHassine/zotero-prime#33]
Here's the error:

ERROR: failed to register layer: failed to Lchown "/opt/code/localstack/localstack/dashboard/web/node_modules/atlassian-soy-cli/node_modules/winston" for UID 3495090557, GID 18285364 (try increasing the number of subordinate IDs in /etc/subuid and /etc/subgid): lchown /opt/code/localstack/localstack/dashboard/web/node_modules/atlassian-soy-cli/node_modules/winston: invalid argument

git pulled >> docker-compose up -d, as recommended in the instructions. Just for fun, tried increasing the UID / GID to let that insane figure be in range, the issue persists with the same error message.

System in question was an unpriveleged LXC on proxmox, fully updated npm / docker-compose, Ubuntu 22.04 LXC Template

Additionally, attempting to follow the non-docker instructions result in a list of mostly deprecated packages that also errors out following npm install:

npm ERR! code 1
npm ERR! path /root/zotero-prime/client/zotero-client/node_modules/node-sass
npm ERR! command failed
npm ERR! command sh -c node scripts/build.js
npm ERR! Building: /usr/bin/node /root/zotero-prime/client/zotero-client/node_modules/node-gyp/bin/node-gyp.js rebuild --verbose --libsass_ext= --libsass_cflags= --libsass_ldflags= --libsass_library=
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp verb cli [
npm ERR! gyp verb cli   '/usr/bin/node',
npm ERR! gyp verb cli   '/root/zotero-prime/client/zotero-client/node_modules/node-gyp/bin/node-gyp.js',
npm ERR! gyp verb cli   'rebuild',
npm ERR! gyp verb cli   '--verbose',
npm ERR! gyp verb cli   '--libsass_ext=',
npm ERR! gyp verb cli   '--libsass_cflags=',
npm ERR! gyp verb cli   '--libsass_ldflags=',
npm ERR! gyp verb cli   '--libsass_library='
npm ERR! gyp verb cli ]
npm ERR! gyp info using node-gyp@3.7.0
npm ERR! gyp info using node@12.22.9 | linux | x64
npm ERR! gyp verb command rebuild []
npm ERR! gyp verb command clean []
npm ERR! gyp verb clean removing "build" directory
npm ERR! gyp verb command configure []
npm ERR! gyp verb check python checking for Python executable "python2" in the PATH
npm ERR! gyp verb `which` failed Error: not found: python2
npm ERR! gyp verb `which` failed     at getNotFoundError (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:13:12)
npm ERR! gyp verb `which` failed     at F (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:68:19)
npm ERR! gyp verb `which` failed     at E (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:80:29)
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/which/which.js:89:16
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/isexe/index.js:42:5
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/isexe/mode.js:8:5
npm ERR! gyp verb `which` failed     at FSReqCallback.oncomplete (fs.js:168:21)
npm ERR! gyp verb `which` failed  python2 Error: not found: python2
npm ERR! gyp verb `which` failed     at getNotFoundError (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:13:12)
npm ERR! gyp verb `which` failed     at F (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:68:19)
npm ERR! gyp verb `which` failed     at E (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:80:29)
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/which/which.js:89:16
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/isexe/index.js:42:5
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/isexe/mode.js:8:5
npm ERR! gyp verb `which` failed     at FSReqCallback.oncomplete (fs.js:168:21) {
npm ERR! gyp verb `which` failed   code: 'ENOENT'
npm ERR! gyp verb `which` failed }
npm ERR! gyp verb check python checking for Python executable "python" in the PATH
npm ERR! gyp verb `which` failed Error: not found: python
npm ERR! gyp verb `which` failed     at getNotFoundError (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:13:12)
npm ERR! gyp verb `which` failed     at F (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:68:19)
npm ERR! gyp verb `which` failed     at E (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:80:29)
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/which/which.js:89:16
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/isexe/index.js:42:5
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/isexe/mode.js:8:5
npm ERR! gyp verb `which` failed     at FSReqCallback.oncomplete (fs.js:168:21)
npm ERR! gyp verb `which` failed  python Error: not found: python
npm ERR! gyp verb `which` failed     at getNotFoundError (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:13:12)
npm ERR! gyp verb `which` failed     at F (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:68:19)
npm ERR! gyp verb `which` failed     at E (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:80:29)
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/which/which.js:89:16
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/isexe/index.js:42:5
npm ERR! gyp verb `which` failed     at /root/zotero-prime/client/zotero-client/node_modules/isexe/mode.js:8:5
npm ERR! gyp verb `which` failed     at FSReqCallback.oncomplete (fs.js:168:21) {
npm ERR! gyp verb `which` failed   code: 'ENOENT'
npm ERR! gyp verb `which` failed }
npm ERR! gyp ERR! configure error 
npm ERR! gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
npm ERR! gyp ERR! stack     at PythonFinder.failNoPython (/root/zotero-prime/client/zotero-client/node_modules/node-gyp/lib/configure.js:492:19)
npm ERR! gyp ERR! stack     at PythonFinder.<anonymous> (/root/zotero-prime/client/zotero-client/node_modules/node-gyp/lib/configure.js:406:16)
npm ERR! gyp ERR! stack     at F (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:68:16)
npm ERR! gyp ERR! stack     at E (/root/zotero-prime/client/zotero-client/node_modules/which/which.js:80:29)
npm ERR! gyp ERR! stack     at /root/zotero-prime/client/zotero-client/node_modules/which/which.js:89:16
npm ERR! gyp ERR! stack     at /root/zotero-prime/client/zotero-client/node_modules/isexe/index.js:42:5
npm ERR! gyp ERR! stack     at /root/zotero-prime/client/zotero-client/node_modules/isexe/mode.js:8:5
npm ERR! gyp ERR! stack     at FSReqCallback.oncomplete (fs.js:168:21)
npm ERR! gyp ERR! System Linux 6.5.11-8-pve
npm ERR! gyp ERR! command "/usr/bin/node" "/root/zotero-prime/client/zotero-client/node_modules/node-gyp/bin/node-gyp.js" "rebuild" "--verbose" "--libsass_ext=" "--libsass_cflags=" "--libsass_ldflags=" "--libsass_library="
npm ERR! gyp ERR! cwd /root/zotero-prime/client/zotero-client/node_modules/node-sass
npm ERR! gyp ERR! node -v v12.22.9
npm ERR! gyp ERR! node-gyp -v v3.7.0
npm ERR! gyp ERR! not ok 
npm ERR! Build failed with error code: 1

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2024-02-21T08_02_44_737Z-debug-0.log

**Comment by anthonyattacks** [comment]
Alright, to document my ongoing efforts to make this work for others:
-First, lxc appears to be a bust. I have no doubt that someone smarter than me can make it work, but up to now, I cannot.
-second, using a VM instead, specifically one based on Ubuntu 18.04 with modern git, docker, docker-compose installed, one can force the pull script to use https:// instead of the deprecated git:// when using --recursive, and that will allow the dependencies to download. Currently, they will time out because git does not support the git:// protocol anymore. I tried updating them manually in config files I found, but that tended to break more than it fixed. Better to hit it at the source. I'll edit this post with the commands I used later for posterity.
-At this point, one can attempt docker-compose up -d from the docker subfolder. This will fail, because the php dependencies are outdated. The best way I've found to get around this up to now is to add a php repository that hosts older php versions and to manually change each mention of php7.0 to php7.2 in the Dockerfile. This is necessary because, at least in my search, the necessary libraries from php7.0 are no longer available. This allows the container to actually build. Will edit to include the repo I used in the future.
-current issue: the init.sh script is failing. It references another script from within the container that, for some reason, cannot communicate to mysql (which, within the container, is fully operational). I've tried updating permissions and pointing it directly at what I think is the root cause, but to no avail as of yet. I'll play with it more this weekend. The container suggests that everything boots without issue, but then the mysql subcontainer terminates and reboots endlessly on a loop. I'll take another crack at it in a day or two and update accordingly.
